[AudioCodecFFMPEG] - disable drc in audio codec context when

Memphiz opened the pull request:
advancedsetting "applydrc" is disabled

This basically hooks up the unused advancedsetting we had laying around to the ffmpeg drc setting.

Thx  to halfgaar from the forum.

ATM the setting defaults to true which means in AC3 streams the DRC as proposed in the stream metadata is applied. This post from halfgaar makes me think that the default should indeed be off:

http://forum.kodi.tv/showthread.php?tid=219228&pid=1965149#pid1965149

Seem this ffmpeg setting only has any effect on AC3 streams if i got it right.

@FernetMenta and @fritsch to decide if we want this and what the default it should have or if we wanna hook this up in the UI or maybe use the "audiooutput.maintainoriginalvolume" gui setting for this...

Memphiz committed:
[AudioCodecFFMPEG] - disable drc in audio codec context when advancedsetting "applydrc" is disabled

fritsch commented:
+1 for the setting but against combining it with the existing setting, cause "original" would be ambiguous. Cause we don't know if original would mean: volume without compression vs. volume with the compression the creator wants us to have - that would confuse users, cause it's not obvious.

Also I am not fully sure what the default value of this boolean should be.

FernetMenta commented:
I like the other pr more: only set drc if advanced setting was changed by user. in all other cases leave the default setting to ffmpeg. if anybody thinks ffmpeg interprets this wrong, please open a thread @ffmpeg-devel

Memphiz commented:
changed it back to only alter if the default advancedsetting is changed ... so i take it the default of this setting (true) is correct - i don't care really ;)

halfgaar commented:
I guess I will take it up with ffmpeg (and then also with libavcodec?), but I really feel strongly the DRC should not be applied by default, and that it in fact is the original without it. Just consider if you have the same film with a DTS soundtrack, or the bluray with AAC. The latter two will have significantly better sound. Why would choosing AC3 as sound source automatically mean the dynamic range should be compressed?

Contrary to Memphiz, I do care :) Very much so, actually. 

DRC was added as an option by Dolby for noisy environments with bad sound systems. It is an alteration of the original stream. It's not called 'compression' without a reason; it compresses the original. Otherwise it would be called dynamic range 'expansion'.

fritsch commented:
@popcornmix it's rather: "use the contained range" or "ignore it". - Adding a double would be another approach to control the drc value directly.

fritsch commented:
@halfgaar a given range is a bit of information, the file has by itself cause the author / creator added that information - so it's used by default. So one also can argument.

Memphiz commented:
@halfgaar i think its best to take this to ffmpeg and ask them why they decided to have this drc turned on by default. If they change there default - we will do so automafically too (and i invert this PR at the same time of course ...)

FernetMenta commented:
also note that ac3 is used by tv broadcasts. a very common scenario is that it gets downmixed and users keep original volume. without drc applied it is more likely to clip. default should be the best setting for the average which is not users having high quality sound systems. most ac3 streams users listen to are tv shows with common stereo equipment. this is not the case for dts.

fritsch commented:
@Memphiz wait a mo. We could add that param as double and default to 0.0 or 1.0 depending on what we decide and make everyone (including the "I want it custom" people happy).

popcornmix commented:
@fritsch but it's an AV_OPT_TYPE_FLOAT:
https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/ac3dec_float.c#L35

Accessing this with av_opt_set_int seems wrong...

Memphiz committed:
[advancedsettings] - changed applydrc to be a float instead of bool

Memphiz commented:
changed it to use av_opt_set_double to set the value and added a second commit which changes the advancedsetting to be a float which defaults to -1.0. If the setting is >= 0.0 i set this value in ffmpeg ...

fritsch commented:
@popcornmix you are obviously right.

FernetMenta commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -109,7 +109,7 @@ void CAdvancedSettings::Initialize()
   m_audioHeadRoom = 0;
   m_ac3Gain = 12.0f;
-  m_audioApplyDrc = true;
+  m_audioApplyDrc = -1.0f;
```
I am wondering if we should promote this to a gui setting

halfgaar replied:
I think so :)

wsnipex replied:
+1

Memphiz replied:
How - those values are really hard to communicate - also this is ffmpeg ac3 specific and most likely can't be used for anything else.

FernetMenta replied:
we provide various levels for that purpose, don't we? but we can do this in a separate pr. no need to block this one

Memphiz commented:
Ok in the sum of the comments i think this PR is good for a start as is? jenkins build this please

Memphiz commented:
stupid typo snucked in again jenkins build this please

Memphiz commented:
Not quiet sure - is there a merge window starting tomorrow for something like this PR or not? @MartijnKaijser

MartijnKaijser commented:
jep.

Memphiz commented:
@NedScott could you adapt the wiki maybe? This value is no bool anymore but a float valuewhere 1.0 means drc applied as suggested from the recorder of the ac3 track and 0.0 means no drc. This also doesn't affect dts but only ac3.

NedScott commented:
Will do.
